### PR TITLE
AL/MIN-475. Fix accrue logic, renaming.

### PR DESCRIPTION
**Pull request by lut1k**
**Description:**

Removed the function `fn calculate_current_pool_data ()`. Implemented the operation of the protocol through calls to the `fn accrue_interest_rate ()` function.
In the pallet, `controller` renamed the variables according to the naming convention.
Changed the public API in the `liquidity-pools` pallet.
Removed pallet `example`.

**Changes in storage or API:**

Changed the structs in the storage:
```
"Pool": {
    "borrowed": "Balance",
    "borrow_index": "Rate",
    "protocol_interest": "Balance"
}
```

```
"PoolUserData": {
    "borrowed": "Balance",
    "interest_index": "Rate",
    "is_collateral": "bool",
    "liquidation_attempts": "u8"
},
````

**Checklist**

- [x] Code is covered with tests
- [x] Code is covered with benchmarks
- [x] Fresh benchmark results included

### CODE REVIEW

File: pallets/traits/src/lib.rs [hunk 1]
```diff
@@ -33,24 +33,94 @@ pub trait PoolsManager<AccountId> {
...
+
+	/// Removes pool data.
+	fn remove_pool_data(pool_id: CurrencyId);
+}
+
+/// Provides functionality for working with a user's storage. Set parameters in storage,
+/// get parameters, check parameters.
+pub trait UserStorageProvider<AccountId, PoolUserData> {
+	/// Sets pool user data.
+	fn set_pool_user_data(who: &AccountId, pool_id: CurrencyId, pool_user_data: PoolUserData);
```
Comment (dev-vadim) [inline pallets/traits/src/lib.rs:78]: According to our Naming convention this should be:
`fn set_user_data(who: &AccountId, pool_id: CurrencyId, user_data: PoolUserData);`

File: pallets/traits/src/lib.rs [hunk 2]
```diff
@@ -33,24 +33,94 @@ pub trait PoolsManager<AccountId> {
...
+	/// Sets the total borrowed and interest index for user.
+	fn set_user_borrow_and_interest_index(
+		who: &AccountId,
+		pool_id: CurrencyId,
+		new_borrow_underlying: Balance,
+		new_interest_index: Rate,
+	);
+
+	/// Gets pool user data.
+	fn get_pool_user_data(pool_id: CurrencyId, who: &AccountId) -> PoolUserData;
```
Comment (dev-vadim) [inline pallets/traits/src/lib.rs:89]: According to our Naming convention:
get_pool_user_data -> get_user_data

File: pallets/controller/rpc/src/lib.rs [hunk 3]
```diff
@@ -38,7 +38,7 @@ pub trait ControllerRpcApi<BlockHash, AccountId> {
 	/// Return:
 	/// - exchange_rate: the Exchange Rate between an mToken and the underlying asset.
 	///	is equal to:
-	/// 	exchange_rate = (total_cash + total_borrowed - total_protocol_interest) / total_supply;
+	/// 	exchange_rate = (pool_cash + pool_borrowed - pool_protocol_interest) / pool_supply;
 	/// 	- borrow_rate: Borrow Interest Rate
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/rpc/src/lib.rs:42]: Fixed.

File: pallets/traits/src/lib.rs (hunk 1 above)
Comment (lut1k) [reply pallets/traits/src/lib.rs:78]: Fixed

File: pallets/traits/src/lib.rs (hunk 2 above)
Comment (lut1k) [reply pallets/traits/src/lib.rs:89]: Fixed

File: pallets/liquidity-pools/src/tests.rs [hunk 4]
```diff
@@ -297,13 +311,13 @@ fn calculate_exchange_rate_should_work() {
 			Ok(Rate::saturating_from_rational(1, 1))
 		);
 
-		// Overflow in calculation: total_cash + total_borrowed
+		// Overflow in calculation: pool_cash + pool_borrowed
```
Comment (kurotych) [inline pallets/liquidity-pools/src/tests.rs]: `pool_cash` should be renamed

File: pallets/traits/src/lib.rs [hunk 5]
```diff
@@ -33,24 +33,94 @@ pub trait PoolsManager<AccountId> {
...
+
+	/// Checks if the user has the collateral.
+	fn check_user_has_collateral(who: &AccountId) -> bool;
+
+	/// Changes the parameter liquidation_attempts depending on the type of liquidation.
+	///
+	/// - `liquidated_pool_id`: the CurrencyId of the pool with loan, for which automatic.
+	/// - `borrower`: the borrower in automatic liquidation.
+	/// - `is_partial_liquidation`: partial or complete liquidation.
+	fn mutate_user_liquidation_attempts(pool_id: CurrencyId, who: &AccountId, is_partial_liquidation: bool);
```
Comment (kurotych) [inline pallets/traits/src/lib.rs:117]: I suggest to split this function to:
`increase_user_liquidation_attempts` and `reset_user_liquidation_attempts`

File: pallets/liquidity-pools/src/tests.rs [hunk 6]
```diff
@@ -241,45 +248,52 @@ fn update_state_on_repay_should_work() {
...
+		.pool_user_data_with_params(DOT, BOB, 0, Rate::default(), true, 0)
+		.pool_user_data_with_params(DOT, CHARLIE, 100, Rate::default(), true, 0)
+		.pool_user_data_with_params(BTC, ALICE, 0, Rate::default(), true, 0)
+		.pool_user_data_with_params(BTC, BOB, 0, Rate::default(), true, 0)
+		.pool_user_data_with_params(BTC, CHARLIE, ONE_HUNDRED, Rate::default(), true, 0)
 		.build()
 		.execute_with(|| {
-			// underlying_amount = 10 * 0.4 = 4
-			assert_eq!(TestPools::convert_from_wrapped(MDOT, 10), Ok(4));
+			assert_eq!(TestPools::get_pool_members_with_loans(DOT), Ok(vec![3, 1]));
```
Comment (kurotych) [inline pallets/liquidity-pools/src/tests.rs:271]: Use constants instead [3, 1]

File: pallets/liquidity-pools/src/lib.rs [hunk 7]
```diff
@@ -525,40 +401,144 @@ impl<T: Config> PoolsManager<T::AccountId> for Pallet<T> {
...
 		Ok(())
 	}
+
+	fn remove_pool_data(pool_id: CurrencyId) {
+		Pools::<T>::remove(pool_id)
+	}
+}
+
+impl<T: Config> CurrencyConverter for Pallet<T> {
+	/// Gets the exchange rate between a mToken and the underlying asset.
```
[the remark this answers is not in the document]
Comment (kurotych) [reply pallets/liquidity-pools/src/lib.rs:467]: Could you add a short math description to comments? (How it calculates)

File: pallets/minterest-protocol/src/lib.rs [hunk 8]
```diff
@@ -582,20 +583,35 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
-	fn do_deposit(who: &T::AccountId, underlying_asset: CurrencyId, underlying_amount: Balance) -> TokensResult {
+	/// Performs the necessary checks for for the existence of currency, check the user's
```
Comment (kurotych) [inline pallets/minterest-protocol/src/lib.rs:586]: > for for

File: pallets/minterest-protocol/src/lib.rs [hunk 9]
```diff
@@ -582,20 +583,35 @@ impl<T: Config> Pallet<T> {
...
+	/// balance, calls `accrue_interest_rate`, `update_mnt_supply_index`, `distribute_supplier_mnt`.
+	/// Transfers an asset into the protocol. The user receives a quantity of mTokens equal
```
Comment (kurotych) [inline pallets/minterest-protocol/src/lib.rs:588]: mTokens => wrapped tokens

File: pallets/controller/rpc/src/lib.rs [hunk 10]
```diff
@@ -38,7 +38,7 @@ pub trait ControllerRpcApi<BlockHash, AccountId> {
 	/// Return:
 	/// - exchange_rate: the Exchange Rate between an mToken and the underlying asset.
 	///	is equal to:
-	/// 	exchange_rate = (total_cash + total_borrowed - total_protocol_interest) / total_supply;
+	/// 	exchange_rate = (pool_cash + pool_borrowed - pool_protocol_interest) / pool_supply;
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/rpc/src/lib.rs]: Fixed

File: pallets/controller/src/lib.rs [hunk 11]
```diff
@@ -1136,17 +1021,18 @@ impl<T: Config> ControllerManager<T::AccountId> for Pallet<T> {
...
 				.ok_or(Error::<T>::NumOverflow)?;
 
-			if <LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) {
-				let m_token_balance = T::MultiCurrency::free_balance(asset, account);
+			if T::LiquidityPoolsManager::check_user_available_collateral(&account, underlying_asset) {
+				let user_supply_wrap = T::MultiCurrency::free_balance(asset, account);
 
 				// sum_collateral += tokens_to_denom * m_token_balance
-				sum_collateral = sum_with_mult_result(sum_collateral, m_token_balance, tokens_to_denom)
+				user_total_collateral = sum_with_mult_result(user_total_collateral, user_supply_wrap, tokens_to_denom)
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:1028]: Fixed

File: pallets/liquidity-pools/src/lib.rs [hunk 12]
```diff
@@ -169,229 +180,74 @@ pub mod module {
...
-
-		Ok(current_exchange_rate)
-	}
-
 	/// Calculates the exchange rate from the underlying to the mToken.
-	/// - `total_cash`: The total amount of underlying tokens the pool has.
-	/// - `total_supply`: Total number of wrapped tokens in circulation.
-	/// - `total_protocol_interest`: Total amount of interest of the underlying held in the pool.
-	/// - `total_borrowed`: Total amount of outstanding borrows of the underlying in this pool.
+	/// - `pool_cash`: The total amount of underlying tokens the pool has.
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidity-pools/src/lib.rs]: FIxed

File: pallets/liquidity-pools/src/lib.rs [hunk 13]
```diff
@@ -421,8 +275,11 @@ impl<T: Config> Pallet<T> {
 		Ok(pools.iter().map(|pool| pool.0).collect::<Vec<CurrencyId>>())
 	}
 
-	/// Checks if the user has the collateral.
-	pub fn check_user_has_collateral(who: &T::AccountId) -> bool {
+	fn check_user_available_collateral(who: &T::AccountId, pool_id: CurrencyId) -> bool {
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidity-pools/src/lib.rs:280]: Fixed

File: pallets/traits/src/lib.rs [hunk 14]
```diff
@@ -33,24 +33,94 @@ pub trait PoolsManager<AccountId> {
...
+
+	/// Sets the parameter `is_collateral` to `true`.
+	fn enable_is_collateral_internal(who: &AccountId, pool_id: CurrencyId);
+
+	/// Sets the parameter `is_collateral` to `false`.
+	fn disable_is_collateral_internal(who: &AccountId, pool_id: CurrencyId);
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/traits/src/lib.rs:123]: Fixed

File: pallets/liquidity-pools/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidity-pools/src/lib.rs:467]: Added

File: pallets/liquidity-pools/src/lib.rs [hunk 15]
```diff
@@ -525,40 +401,144 @@ impl<T: Config> PoolsManager<T::AccountId> for Pallet<T> {
...
+	fn usd_to_underlying(usd_amount: Balance, oracle_price: Price) -> Result<Balance, DispatchError> {
+		let underlying_amount = Rate::from_inner(usd_amount)
+			.checked_div(&oracle_price)
+			.map(|x| x.into_inner())
+			.ok_or(Error::<T>::ConversionError)?;
+		Ok(underlying_amount)
+	}
+
+	/// Converts a specified amount of USD into wrapped tokens.
+	fn usd_to_wrapped(usd_amount: Balance, exchange_rate: Rate, oracle_price: Price) -> Result<Balance, DispatchError> {
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidity-pools/src/lib.rs:538]: Fixed

File: pallets/liquidity-pools/src/tests.rs (hunk 4 above)
Comment (lut1k) [reply pallets/liquidity-pools/src/tests.rs]: Renamed

File: pallets/traits/src/lib.rs (hunk 5 above)
Comment (lut1k) [reply pallets/traits/src/lib.rs:117]: Splitted

File: pallets/liquidity-pools/src/tests.rs (hunk 6 above)
Comment (lut1k) [reply pallets/liquidity-pools/src/tests.rs:271]: Fixed

File: pallets/minterest-protocol/src/lib.rs (hunk 8 above)
Comment (lut1k) [reply pallets/minterest-protocol/src/lib.rs:586]: Fixed

File: pallets/minterest-protocol/src/lib.rs (hunk 9 above)
Comment (lut1k) [reply pallets/minterest-protocol/src/lib.rs:588]: Fixed

File: pallets/controller/src/lib.rs [hunk 16]
```diff
@@ -1021,10 +1021,10 @@ impl<T: Config> ControllerManager<T::AccountId> for Pallet<T> {
 				.and_then(|v| v.checked_mul(&oracle_price))
 				.ok_or(Error::<T>::NumOverflow)?;
 
-			if T::LiquidityPoolsManager::check_user_available_collateral(&account, underlying_asset) {
+			if T::LiquidityPoolsManager::is_pool_collateral(&account, underlying_asset) {
 				let user_supply_wrap = T::MultiCurrency::free_balance(asset, account);
 
-				// sum_collateral += tokens_to_denom * m_token_balance
+				// user_total_collateral += tokens_to_denom * m_token_balance
```
Comment (kurotych) [inline pallets/controller/src/lib.rs:1027]: m_token_balance should be renamed

**Review by kurotych — APPROVED** [review]

File: pallets/controller/src/lib.rs (hunk 16 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs]: Renamed

File: pallets/controller/src/lib.rs [hunk 17]
```diff
@@ -880,26 +843,27 @@ impl<T: Config> Pallet<T> {
...
 	/// - `pool_supply_underlying_balance`: The amount of cash in the pool.
-	/// - `pool_borrowed_balance`: The amount of borrows in the pool.
+	/// - `pool_borrow_underlying`: The amount of borrows in the pool.
 	/// - `pool_protocol_interest`: The amount of interest in the pool (currently unused).
 	///
-	/// returns `utilization_rate =
-	///  pool_borrows / (pool_cash + pool_borrows - pool_protocol_interest)`
+	/// returns `utilization_rate = pool_borrow_underlying /
+	/// (pool_supply_underlying_balance + pool_borrow_underlying - pool_protocol_interest)`
 	fn calculate_utilization_rate(
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:851]: Updated

**Review by kurotych — APPROVED** [review]

File: pallets/controller/src/lib.rs [hunk 18]
```diff
@@ -479,18 +486,14 @@ impl<T: Config> Pallet<T> {
...
-		.ok()
+		Self::accrue_interest_rate(pool_id).ok()?;
+		let pool_supply_underlying = T::LiquidityPoolsManager::get_pool_available_liquidity(pool_id);
+		let pool_data = T::LiquidityPoolsManager::get_pool_data(pool_id);
+		Self::calculate_utilization_rate(pool_supply_underlying, pool_data.borrowed, pool_data.protocol_interest).ok()
 	}
 
 	/// Calculates user total supply and user total borrowed balance in usd based on
-	/// total_borrowed, total_protocol_interest, borrow_index values calculated for current block.
+	/// pool_borrowed, pool_protocol_interest, borrow_index values calculated for current block.
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:496]: FIxed

File: pallets/controller/src/lib.rs [hunk 19]
```diff
@@ -508,139 +511,109 @@ impl<T: Config> Pallet<T> {
...
-	///     `fresh_pool_protocol_interest` - freshest value of protocol interest in the n pool;
+	///     `fresh_pool_protocol_interest_usd` - freshest value of protocol interest in the n pool;
 	pub fn get_protocol_total_values() -> result::Result<(Balance, Balance, Balance, Balance), DispatchError> {
 		CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
 			.iter()
 			.filter(|&underlying_id| T::LiquidityPoolsManager::pool_exists(underlying_id))
 			.try_fold(
 				(Balance::zero(), Balance::zero(), Balance::zero(), Balance::zero()),
-				|(protocol_available_liquidity, protocol_borrow, tvl, protocol_interest),
+				|(protocol_supply_usd, protocol_borrow_usd, tvl, protocol_interest_usd),
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:562]: Fixed

File: pallets/liquidation-pools/src/lib.rs [hunk 20]
```diff
@@ -64,6 +65,7 @@ type BalanceResult = sp_std::result::Result<Balance, DispatchError>;
 #[frame_support::pallet]
 pub mod module {
 	use super::*;
+	use pallet_traits::CurrencyConverter;
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs:68]: Moved

File: pallets/liquidity-pools/src/tests.rs [hunk 21]
```diff
@@ -297,13 +311,13 @@ fn calculate_exchange_rate_should_work() {
...
 
-		// Overflow in calculation: total_cash + total_borrowed
+		// Overflow in calculation: pool_supply_underlying + pool_borrow_underlying
 		assert_noop!(
 			TestPools::calculate_exchange_rate(Balance::max_value(), 100, 100, 100),
 			Error::<Test>::ExchangeRateCalculationError
 		);
 
-		// Overflow in calculation: cash_plus_borrows - total_protocol_interest
+		// Overflow in calculation: cash_plus_borrows - pool_protocol_interest
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidity-pools/src/tests.rs:320]: Fixed

File: pallets/mnt-token/src/lib.rs [hunk 22]
```diff
@@ -545,12 +547,12 @@ impl<T: Config> MntManager<T::AccountId> for Pallet<T> {
...
 
-		let total_cash = T::LiquidityPoolsManager::get_pool_available_liquidity(pool_id);
-		let total_protocol_interest = T::LiquidityPoolsManager::get_pool_total_protocol_interest(pool_id);
+		let pool_supply_underlying = T::LiquidityPoolsManager::get_pool_available_liquidity(pool_id);
+		let pool_protocol_interest = T::LiquidityPoolsManager::get_pool_protocol_interest(pool_id);
 
-		let total_supply = total_cash
-			.checked_sub(total_protocol_interest)
-			.and_then(|v| v.checked_add(total_borrow))
+		let pool_supply_wrap = pool_supply_underlying
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/mnt-token/src/lib.rs:553]: Fixed

**Review by kurotych — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**